File: experiments-local/experiment-opencode-easy-final/runs/agent=opencode_language=typescript_model=openrouter/z-ai/glm-5.2_tooling=none/rep1/src/bookStore.ts
import Database from "better-sqlite3";
import type { Database as DatabaseConnection } from "better-sqlite3";
import type { Book, CreateBookInput, UpdateBookInput } from "./types";

const SCHEMA = `
  CREATE TABLE IF NOT EXISTS books (
    id INTEGER PRIMARY KEY AUTOINCREMENT,
    title TEXT NOT NULL,
    author TEXT NOT NULL,
    year INTEGER,
    isbn TEXT
  );
`;

export class BookStore {
  private db: DatabaseConnection;

  constructor(dbPath: string = ":memory:") {
    this.db = new Database(dbPath);
    this.db.pragma("journal_mode = WAL");
    this.db.exec(SCHEMA);
  }

  create(input: CreateBookInput): Book {
    const stmt = this.db.prepare(
      "INSERT INTO books (title, author, year, isbn) VALUES (?, ?, ?, ?)"
    );
    const result = stmt.run(
      input.title,
      input.author,
      input.year ?? null,
      input.isbn ?? null
    );
    return this.getById(Number(result.lastInsertRowid))!;
  }

  list(authorFilter?: string): Book[] {
    if (authorFilter) {
      const stmt = this.db.prepare(
        "SELECT * FROM books WHERE author = ? ORDER BY id"
      );
      return stmt.all(authorFilter) as Book[];
    }
    const stmt = this.db.prepare("SELECT * FROM books ORDER BY id");
    return stmt.all() as Book[];
  }

  getById(id: number): Book | undefined {
    const stmt = this.db.prepare("SELECT * FROM books WHERE id = ?");
    return stmt.get(id) as Book | undefined;
  }

  update(id: number, input: UpdateBookInput): Book | undefined {
    const existing = this.getById(id);
    if (!existing) return undefined;

    const next: Book = {
      ...existing,
      title: input.title ?? existing.title,
      author: input.author ?? existing.author,
      year: input.year ?? existing.year,
      isbn: input.isbn ?? existing.isbn,
    };

    const stmt = this.db.prepare(
      "UPDATE books SET title = ?, author = ?, year = ?, isbn = ? WHERE id = ?"
    );
    stmt.run(next.title, next.author, next.year, next.isbn, id);
    return this.getById(id);
  }

  delete(id: number): boolean {
    const stmt = this.db.prepare("DELETE FROM books WHERE id = ?");
    const result = stmt.run(id);
    return result.changes > 0;
  }

  close(): void {
    this.db.close();
  }
}

File: experiments-local/experiment-15/runs/agent=omp_language=go_model=openrouter/deepseek/deepseek-v3.2_tooling=none/rep2/database/book_store.go
package database

import (
	"database/sql"
	"errors"
	"fmt"
	"log"
	"time"

	"bookapi/models"

	"github.com/google/uuid"
)

type BookStoreDB struct {
	db *sql.DB
}

func NewBookStore(db *sql.DB) *BookStoreDB {
	return &BookStoreDB{db: db}
}

func (s *BookStoreDB) Create(req *models.BookRequest) (*models.Book, error) {
	if req.Title == "" || req.Author == "" {
		return nil, errors.New("title and author are required")
	}

	id := uuid.New().String()
	now := time.Now()

	query := `
		INSERT INTO books (id, title, author, year, isbn, created_at, updated_at)
		VALUES (?, ?, ?, ?, ?, ?, ?)
	`
	_, err := s.db.Exec(query, id, req.Title, req.Author, req.Year, req.ISBN, now, now)
	if err != nil {
		return nil, fmt.Errorf("failed to create book: %w", err)
	}

	book := &models.Book{
		ID:        id,
		Title:     req.Title,
		Author:    req.Author,
		Year:      req.Year,
		ISBN:      req.ISBN,
		CreatedAt: now,
		UpdatedAt: now,
	}

	log.Printf("Created book with ID: %s", id)
	return book, nil
}

func (s *BookStoreDB) GetByID(id string) (*models.Book, error) {
	query := `
		SELECT id, title, author, year, isbn, created_at, updated_at
		FROM books
		WHERE id = ?
	`
	row := s.db.QueryRow(query, id)

	var book models.Book
	err := row.Scan(&book.ID, &book.Title, &book.Author, &book.Year, &book.ISBN, &book.CreatedAt, &book.UpdatedAt)
	if err != nil {
		if errors.Is(err, sql.ErrNoRows) {
			return nil, errors.New("book not found")
		}
		return nil, fmt.Errorf("failed to get book: %w", err)
	}

	return &book, nil
}

func (s *BookStoreDB) List(authorFilter string) ([]models.Book, error) {
	var query string
	var args []interface{}

	if authorFilter != "" {
		query = `
			SELECT id, title, author, year, isbn, created_at, updated_at
			FROM books
			WHERE author LIKE ?
			ORDER BY created_at DESC
		`
		args = append(args, "%"+authorFilter+"%")
	} else {
		query = `
			SELECT id, title, author, year, isbn, created_at, updated_at
			FROM books
			ORDER BY created_at DESC
		`
	}

	rows, err := s.db.Query(query, args...)
	if err != nil {
		return nil, fmt.Errorf("failed to query books: %w", err)
	}
	defer rows.Close()

	var books []models.Book
	for rows.Next() {
		var book models.Book
		err := rows.Scan(&book.ID, &book.Title, &book.Author, &book.Year, &book.ISBN, &book.CreatedAt, &book.UpdatedAt)
		if err != nil {
			return nil, fmt.Errorf("failed to scan book: %w", err)
		}
		books = append(books, book)
	}

	if err := rows.Err(); err != nil {
		return nil, fmt.Errorf("error iterating rows: %w", err)
	}

	return books, nil
}

func (s *BookStoreDB) Update(id string, req *models.BookRequest) (*models.Book, error) {
	if req.Title == "" || req.Author == "" {
		return nil, errors.New("title and author are required")
	}

	// Check if book exists
	existing, err := s.GetByID(id)
	if err != nil {
		return nil, err
	}

	now := time.Now()
	query := `
		UPDATE books
		SET title = ?, author = ?, year = ?, isbn = ?, updated_at = ?
		WHERE id = ?
	`
	_, err = s.db.Exec(query, req.Title, req.Author, req.Year, req.ISBN, now, id)
	if err != nil {
		return nil, fmt.Errorf("failed to update book: %w", err)
	}

	updatedBook := &models.Book{
		ID:        id,
		Title:     req.Title,
		Author:    req.Author,
		Year:      req.Year,
		ISBN:      req.ISBN,
		CreatedAt: existing.CreatedAt,
		UpdatedAt: now,
	}

	log.Printf("Updated book with ID: %s", id)
	return updatedBook, nil
}

func (s *BookStoreDB) Delete(id string) error {
	// Check if book exists
	_, err := s.GetByID(id)
	if err != nil {
		return err
	}

	query := "DELETE FROM books WHERE id = ?"
	result, err := s.db.Exec(query, id)
	if err != nil {
		return fmt.Errorf("failed to delete book: %w", err)
	}

	rowsAffected, err := result.RowsAffected()
	if err != nil {
		return fmt.Errorf("failed to get rows affected: %w", err)
	}

	if rowsAffected == 0 {
		return errors.New("book not found")
	}

	log.Printf("Deleted book with ID: %s", id)
	return nil
}